File: gitea_repos_master/whatsapp/whatsapp-baileys-api/src/logic/mongo/connection.ts
import { Collection, Db, MongoClient } from 'mongodb';
import { MongoFields } from './fields';

export class MongoConnection extends MongoFields {
  private client: MongoClient;
  private db: Db;
  public sessions: Collection;

  constructor(mongoURL: string, db: string, collection: string | null) {
    super();

    this.client = new MongoClient(mongoURL, {
      connectTimeoutMS: 8_000,
      socketTimeoutMS: 8_000,
      maxConnecting: 3,
      serverSelectionTimeoutMS: 5_000
    });
    this.db = this.client.db(db);
    this.sessions = collection ? this.db.collection(collection) : null;
  }

  public async connect(): Promise<MongoConnection> {
    await this.client.connect();
    await this._preprocessCollection();
    return this;
  }

  public async close(): Promise<void> {
    await this.client.close();
  }

  public switchCollection(collection: string): void {
    this.sessions = this.db.collection(collection);
  }

  protected async _preprocessCollection(): Promise<void> {
    await this._createIndexes();
  }

  private async _createIndexes(): Promise<void> {
    if (this.sessions === null) {
      return;
    }
    const indexes = await this.sessions.indexes();
    if (Object.keys(indexes).length === 1) {
      await this.sessions.createIndex([{ lastUse: 1 }]);
    }
  }
}

File: gitea_repos_master/isphere-classes/mongo-client/mongo_client/connection.py
from mongo_client.fields import MongoFields
from mongo_client.logger import logger
from mongo_client.operations import MongoOperations


class MongoConnectionInterface(MongoOperations, MongoFields):
    """Класс коннектор к БД"""

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

    async def connect(self):
        await super().connect()
        await self._preprocess_collection()
        return self

    def switch_collection(self, collection: str):
        self.sessions = self.db[collection]

    async def _preprocess_collection(self):
        await self._create_indexes()

    async def _create_indexes(self):
        """Функция создания индексов по умолчанию"""
        if self.sessions is None:
            return

        own_indexes = {"last_use_1": [(self.last_use, 1)]}

        indexes = await self.index_information()

        current_indexes = list(indexes.keys())
        allowed_indexes = ["_id_"] + list(own_indexes.keys())

        excess_indexes = set(current_indexes) - set(allowed_indexes)
        for ind in excess_indexes:
            await self.drop_index(ind)

        for ind, value in own_indexes.items():
            if ind not in current_indexes:
                await self.create_index(value)
                logger.debug(f"Mongo created new index: {value}")
